### PR TITLE
Fix terribly broken, fragile and potentially cyclic linking

### PR DESCRIPTION
**Pull request by rschu1ze**
Sorry for the clickbait title. This is about static method `ConnectionTimeouts::getHTTPTimeouts()`. It is declared in header IO/ConnectionTimeouts.h, and defined in header IO/ConnectionTimeoutsContext.h. This is weird and caused issues with linking on s390x (#45520). There was an attempt to fix some inconsistencies (#45848) but neither did @Algunenano nor me at first really understood why the definition is in the header.

Turns out that ConnectionTimeoutsContext.h is only #include'd from source files which are part of the normal server build BUT NOT part of the keeper standalone build (which must be enabled via CMake `-DBUILD_STANDALONE_KEEPER=1`). This dependency was not documented and as a result, some misguided workarounds were introduced earlier, e.g. https://github.com/ClickHouse/ClickHouse/commit/0341c6c54bd7ac77200b4ca123208b195514ef20

The deeper cause was that `getHTTPTimeouts()` is passed a `Context`. This class is part of the "dbms" library which is deliberately not linked by the standalone build of clickhouse-keeper. `Context` also cannot be build into clickhouse-keeper because it has too many transitive (source) dependencies. The context is only used to get `Settings` which are part of the clickhouse_common library and which is linked by clickhouse-keeper already.

To resolve this mess, this PR

- creates source file IO/ConnectionTimeouts.cpp and moves all ConnectionTimeouts definitions into it, including `getHTTPTimeouts()`.

- breaks the wrong dependency by passing `Settings` instead of `Context` into `getHTTPTimeouts()`. A little bit of logic is moved to the caller side.

- removes the previous hacks. There is actually another hack in "Core/SettingsFields.cpp" which cannot be resolved easily and was documented for now.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by Algunenano** [comment]
Is this already covered by https://github.com/ClickHouse/ClickHouse/pull/45848 ??

**Comment by rschu1ze** [comment]
Interesting that you found the same problem at the same time.

I think your other PR is okay but actually spreading the three static methods across two headers is silly (and that is not addressed in your PR). I suspect your PR does not fix the s390x linker issue (see PR description), but I am rather sure that this PR will fix or at least uncover further details of the issue. So I think this PR still makes sense to have.,

**Comment by rschu1ze** [comment]
ClickHouse Integration Tests (tsan) [6/6]: #46056

ClickHouse Stateless Tests (asan) [3/4]:
- 02346_full_text_search: This tests the inverted index and it seems a bit flaky when accessing system.query_log. I think there is sometimes parallel activity which messes with the expected results. Can look into this separately, but it is for sure unrelated.

**Comment by alexey-milovidov** [comment]
Let's fix the `02346_full_text_search` test before merging this PR.

File: src/BridgeHelper/XDBCBridgeHelper.h [hunk 1]
```diff
@@ -96,9 +96,14 @@ class XDBCBridgeHelper : public IXDBCBridgeHelper
 
     bool bridgeHandShake() override
     {
+        const auto & settings = getContext()->getSettingsRef();
+        const auto & cfg = getContext()->getConfigRef();
+        Poco::Timespan http_keep_alive_timeout{cfg.getUInt("keep_alive_timeout", 10), 0};
+        auto timeouts = ConnectionTimeouts::getHTTPTimeouts(settings, http_keep_alive_timeout);
```
Comment (serxa) [inline src/BridgeHelper/XDBCBridgeHelper.h:102]: It feels really ugly to repeat this code over and over... Is there still a way to do it in some common place? Or at least hide the constant?

Comment (rschu1ze) [reply src/BridgeHelper/XDBCBridgeHelper.h:102]: Agree, will make this less verbose.

**Review by serxa — APPROVED** [review]

**Comment by rschu1ze** [comment]
> Let's fix the `02346_full_text_search` test before merging this PR.

About 02346_full_text_search: It fails sporadically, I try to stabilize it right now. Will make a new PR for this as the failures are unrelated to this PR.

**Pull request merged**